### DISCUSSION TITLE
Blue-green deployment

### DISCUSSION THREAD

**Discussion by mjulianotq** [afs/rdf-delta#60] [Q&A] [answered]
Does RDF Delta lend itself well to blue-green deployments? If I understand it correctly, as long as the storage format stays relatively stable, it shouldn't be a problem. The only time I would imagine it would be a problem is if there is a unidirectional migration of the underlying data that forces you to restore from a backup in the event you need to revert. Does that sound correct? If so, is there much risk of such unidirectional migrations coming in future versions (post 0.9.1)?

**Comment by afs** [comment, marked as answer]
Could you describe the setup you are working with? Single patch server or replicated patch server?

From the description you want to share old patches - are you asking about blue-green deployment of just the application server using RDF-delta or the whole setup top to bottom?

The 0.9.x release line is functionality and disk format stable and will become 1.0.0.

**Reply by mjulianotq** [reply]
I have not set it up yet. I'm trying to anticipate future updates. It will be a replicated patch server. I would want to perform the deployment piece-wise, so I'd probably start with the patch servers since they're much lighter weight than the application using them.

**Reply by afs** [reply]
The storage of patches is coordinated by the patch servers. The index of patches is in the patch servers, not the patch storage (which is basically blob store)

So you can't do a blue-green setup of patch server clusters because there is state that will get out of step just like if you tried blue-green between SQL database instances.

If you halt updates and carefully clone the system it wil work; if you get it wrong, the system is corrupt and will have to be rebuilt. i.e it's high-risk.

But you can do a rolling upgrade. Simply kill a patch server, and start a new one with newer software; move on to the next server. The single failure is handled by the system.

**Reply by mjulianotq** [reply]
A rolling update should suit my needs. Thanks, Andy!
